powerstore/resource_protection_policy.go

```diff
@@ -296,29 +328,103 @@ func (r resourceProtectionPolicy) updatePolicyState(polState *models.ProtectionP
 	var snapshotRuleIds []string
 	for _, snapshotRule := range polResponse.SnapshotRules {
 		snapshotRuleIds = append(snapshotRuleIds, snapshotRule.ID)
 	}
-	snapshotList := []attr.Value{}
+	snapshotIDList := []attr.Value{}
 	for _, snapshotRuleID := range snapshotRuleIds {
-		snapshotList = append(snapshotList, types.String{Value: string(snapshotRuleID)})
+		snapshotIDList = append(snapshotIDList, types.String{Value: string(snapshotRuleID)})
 	}
 	polState.SnapshotRuleIDs = types.Set{
 		ElemType: types.StringType,
-		Elems:    snapshotList,
+		Elems:    snapshotIDList,
+	}
+
+	//Update SnapshotRuleNames value from Plan to State
+	var snapshotRuleNames []string
+	for _, snapshotRuleName := range polPlan.SnapshotRuleNames.Elems {
+		snapshotRuleNames = append(snapshotRuleNames, strings.Trim(snapshotRuleName.String(), "\""))
+	}
+	snapshotNameList := []attr.Value{}
+	for _, snapshotRuleName := range snapshotRuleNames {
+		snapshotNameList = append(snapshotNameList, types.String{Value: string(snapshotRuleName)})
+	}
+	polState.SnapshotRuleNames = types.List{
+		ElemType: types.StringType,
+		Elems:    snapshotNameList,
+	}
+}
+
+func (r resourceProtectionPolicy) fetchByName(plan *models.ProtectionPolicy) (valid bool, err string) {
+	var snapshotRuleIds []string
+	if len(plan.SnapshotRuleIDs.Elems) != 0 && len(plan.SnapshotRuleNames.Elems) != 0 {
+		return false, "Snapshot Rule ID or Snapshot Rule Name"
+	} else if len(plan.SnapshotRuleNames.Elems) != 0 {
+		for _, snapshotRuleName := range plan.SnapshotRuleNames.Elems {
+			snapshotRule, _ := r.p.client.PStoreClient.GetSnapshotRuleByName(context.Background(), strings.Trim(snapshotRuleName.String(), "\""))
+			snapshotRuleIds = append(snapshotRuleIds, strings.Trim(snapshotRule.ID, "\""))
+
+			snapshotList := []attr.Value{}
+			for i := 0; i < len(snapshotRuleIds); i++ {
+				snapshotList = append(snapshotList, types.String{Value: string(snapshotRuleIds[i])})
+			}
+
+			plan.SnapshotRuleIDs = types.Set{
+				ElemType: types.StringType,
+				Elems:    snapshotList,
+			}
+		}
 	}
+
+	var replicationRuleIds []string
+	if len(plan.ReplicationRuleIDs.Elems) != 0 && len(plan.ReplicationRuleNames.Elems) != 0 {
+		return false, "Replication Rule ID or Replication Rule Name"
+	} else if len(plan.ReplicationRuleNames.Elems) != 0 {
+		for _, replicationRuleName := range plan.ReplicationRuleNames.Elems {
+			replicationRule, _ := r.p.client.PStoreClient.GetReplicationRuleByName(context.Background(), strings.Trim(replicationRuleName.String(), "\""))
+			replicationRuleIds = append(replicationRuleIds, strings.Trim(replicationRule.ID, "\""))
+		}
+
+		replicationList := []attr.Value{}
+		for i := 0; i < len(replicationRuleIds); i++ {
+			replicationList = append(replicationList, types.String{Value: string(replicationRuleIds[i])})
+		}
+
+		plan.ReplicationRuleIDs = types.List{
+			ElemType: types.StringType,
+			Elems:    replicationList,
+		}
+	}
+
+	return true, ""
```

Comment:
same as above for error
and as you are using named return, so u may used naked return
https://go.dev/tour/basics/7

Reply:
Thanks for the suggestions. Made changes wherever applicable.